specifyweb/specify/tree_ranks.py

```diff
@@ -0,0 +1,314 @@
+    'storage': (STORAGE_RANKS, STORAGE_RANK_INCREMENT),
+    'geologictimeperiod': (GEOLOGIC_TIME_PERIOD_RANKS, GEOLOGIC_TIME_PERIOD_INCREMENT),
+    'lithostrat': (LITHO_STRAT_RANKS, LITHO_STRAT_INCREMENT),
+}
+
+TREE_RANK_TO_ITEM_MAP = {
+    'Taxontreedefitem': 'Taxon',
+    'Geographytreedefitem': 'Geography',
+    'Storagetreedefitem': 'Storage',
+    'Geologictimeperiodtreedefitem': 'Geologictimeperiod',
+    'Lithostrattreedefitem': 'Lithostrat'
+}
+
+def get_tree_item_model(tree_rank_model_name):
+    tree_item_model_name = TREE_RANK_TO_ITEM_MAP.get(tree_rank_model_name.title(), None)
+    if not tree_item_model_name:
+        return None
+    return getattr(spmodels, tree_item_model_name, None)
+
+def tree_rank_count(tree_rank_model_name, tree_rank_id) -> int:
+    tree_item_model = get_tree_item_model(tree_rank_model_name)
+    if not tree_item_model:
+        return 0
+    return tree_item_model.objects.filter(definitionitem_id=tree_rank_id).count()
+
+def is_tree_rank_empty(tree_rank_model, tree_rank) -> bool:
+    tree_item_model = get_tree_item_model(tree_rank_model.__name__)
+    if not tree_item_model:
+        return False
+    return tree_item_model.objects.filter(definitionitem=tree_rank).count() == 0
+
+def post_tree_rank_save(tree_def_item_model, new_rank):
+    tree_def = new_rank.treedef
+    parent_rank = new_rank.parent
+    new_rank_id = new_rank.rankid
+
+    # Set the parent rank, that previously pointed to the target, to the new rank
+    child_ranks = tree_def_item_model.objects.filter(parent=parent_rank).exclude(id=new_rank.id).update(parent=new_rank)
+
+    # Regenerate full names
+    tree_extras.set_fullnames(tree_def, null_only=False, node_number_range=None)
+
+def pre_tree_rank_deletion(tree_def_item_model, rank):
+    tree_def = rank.treedef
+
+    # Make sure no nodes are present in the rank before deleting rank
+    if tree_def_item_model.objects.filter(parent=rank).count() > 1:
+        raise TreeBusinessRuleException("The Rank {rank.name} is not empty, cannot delete!")
+
+    # Set the parent rank, that previously pointed to the old rank, to the target rank
+    child_ranks = tree_def_item_model.objects.filter(parent=rank).update(parent=rank.parent)
+
+def post_tree_rank_deletion(rank):
+    # Regenerate full names
+    tree_extras.set_fullnames(rank.treedef, null_only=False, node_number_range=None)
+
+def pre_tree_rank_init(new_rank):
+    set_rank_id(new_rank)
+
+def set_rank_id(new_rank):
+    """
+    Sets the new rank to the specified tree when adding a new rank.
+    Expects at least the name, parent, and tree_def of the rank to be set.
+    All the other parameters are optional.
+    """
+    # Get parameter values from data
+    tree = new_rank.specify_model.name.replace("TreeDefItem", "").lower()
+    new_rank_name = getattr(new_rank, 'name', None)
+    parent_rank_name = getattr(new_rank.parent, 'name', 'root') if getattr(new_rank, 'parent', None) else 'root'
```

Comment:
you need to be clear on what you are getting in attr.
```
  new_rank_name = getattr(new_rank, 'name', None)
  parent_rank_name = getattr(new_rank.parent, 'name', 'root') if getattr(new_rank, 'parent', None) else 'root'
```
you don't want that. you want
```
    new_rank_name = new_rank.name
    parent_rank_name = new_rank.parent.name if new_rank.parent else 'root'
```
think why. we know these fields must exist, don't overcomplicate it.